packages/flutter_tools/lib/src/android/apk_locator.dart

```diff
@@ -0,0 +1,137 @@
+    @required AndroidBuildInfo androidBuildInfo,
+  }) {
+    // First check if the tool can find the exact file.
+    final List<File> exactFiles = <File>[];
+    final BuildInfo buildInfo = androidBuildInfo.buildInfo;
+    final String modeName = camelCase(buildInfo.modeName);
+
+    for (final String fileName in _apkFilesNamesFor(androidBuildInfo)) {
+      final File apkFile = apkDirectory
+        .childDirectory(modeName)
+        .childFile(fileName);
+      if (apkFile.existsSync()) {
+        exactFiles.add(apkFile);
+        continue;
+      }
+
+      if (buildInfo.flavor != null) {
+        // Android Studio Gradle plugin v3 adds flavor to path.
+        final File apkFile = apkDirectory
+          .childDirectory(buildInfo.flavor)
+          .childDirectory(modeName)
+          .childFile(fileName);
+        if (apkFile.existsSync()) {
+          exactFiles.add(apkFile);
+          continue;
+        }
+      }
+    }
+
+    // If at least one APK file was found, skip the heuristics.
+    if (exactFiles.isNotEmpty) {
+      return exactFiles;
+    }
+    if (!apkDirectory.existsSync()) {
+      return const <File>[];
+    }
+
+    // If the process has reached this point, the tool was unable to determine
+    // the exact APK name. Apply heursitics to sort more likely candiates towards
+    // the start of the list.
+    bool filter(File file) {
+      // The file is not a .apk
+      if (!_fileSystem.path.basename(file.path).endsWith('.apk')) {
+        return false;
+      }
+      // The file is an exclude file, such as app.apk where the output
+      // is copied to.
+      final String relativePath = _fileSystem.path
+        .relative(file.path, from: apkDirectory.path);
+      if (excludePaths.contains(relativePath)) {
+        return false;
+      }
+      return true;
+    }
+
+    int sort(File a, File b) {
+      // Check whether the apk contains a string for the build mode, either
+      // in the file name or directory path.
+      final String buildModeName = getNameForBuildMode(androidBuildInfo.buildInfo.mode);
+      final bool aContainsBuildMode = a.path
+        .contains(buildModeName);
+      final bool bContainsBuildMode = b.path
+        .contains(buildModeName);
+      if (aContainsBuildMode && !bContainsBuildMode) {
+        return -1;
+      } else if (!aContainsBuildMode && bContainsBuildMode) {
+        return 1;
+      }
+
+      // Check if the flavor name is contained in the apk or path
+      // when a flavor is involved.
+      if (androidBuildInfo.buildInfo.flavor?.isNotEmpty ?? false) {
+        final bool aContainsFlavorName = a.path
+          .contains(androidBuildInfo.buildInfo.flavor);
+        final bool bContainsFlavorName = b.path
+          .contains(androidBuildInfo.buildInfo.flavor);
+        if (aContainsFlavorName && !bContainsFlavorName) {
+          return -1;
+        } else if (!aContainsFlavorName && bContainsFlavorName) {
+          return 1;
+        }
+      }
```

Comment:
What about we just compare the last modified timestamp? That might be sufficient to get the latest APK generated by Gradle.